### PR TITLE
planner: choose a better plan when the range condition can convert to point

### LINKED ISSUE #17153: Can't choose better plan when the range condition can convert to point.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

```
drop table t;
create table t(a int,b int,index(a,b));
insert into t values(1,1);
insert into t values(2,2);
insert into t values(2,2);
insert into t values(2,2);
insert into t values(2,2);
insert into t values(2,2);
insert into t values(3,3);
desc select * from t where a >= 2 and a <= 2 order by b limit 2;
desc select * from t where a = 2 order by b limit 2;
```

### 2. What did you expect to see?
Same plan.
### 3. What did you see instead?
```
tidb> desc select * from t where a >= 2 and a <= 2 order by b limit 2;
+-----------------------------+---------+-----------+------------------------+---------------------------------------------+
| id                          | estRows | task      | access object          | operator info                               |
+-----------------------------+---------+-----------+------------------------+---------------------------------------------+
| TopN_8                      | 2.00    | root      |                        | test.t.b:asc, offset:0, count:2             |
| └─IndexReader_16            | 2.00    | root      |                        | index:TopN_15                               |
|   └─TopN_15                 | 2.00    | cop[tikv] |                        | test.t.b:asc, offset:0, count:2             |
|     └─IndexRangeScan_14     | 10.00   | cop[tikv] | table:t, index:a(a, b) | range:[2,2], keep order:false, stats:pseudo |
+-----------------------------+---------+-----------+------------------------+---------------------------------------------+
4 rows in set (0.00 sec)

tidb> desc select * from t where a = 2 order by b limit 2;
+-----------------------------+---------+-----------+------------------------+--------------------------------------------+
| id                          | estRows | task      | access object          | operator info                              |
+-----------------------------+---------+-----------+------------------------+--------------------------------------------+
| Limit_11                    | 2.00    | root      |                        | offset:0, count:2                          |
| └─IndexReader_21            | 2.00    | root      |                        | index:Limit_20                             |
|   └─Limit_20                | 2.00    | cop[tikv] |                        | offset:0, count:2                          |
|     └─IndexRangeScan_19     | 2.00    | cop[tikv] | table:t, index:a(a, b) | range:[2,2], keep order:true, stats:pseudo |
+-----------------------------+---------+-----------+------------------------+--------------------------------------------+
4 rows in set (0.00 sec)

```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
master

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: close #17153 

Problem Summary:
Can't choose a better plan when the range condition can convert to point.

### What is changed and how it works?
Some expressions such as `a >= 2 && a <= 2` can also convert to a equal condition like `a = 2`. We should take it into consideration. We check whether the range of range is a point to determine whether it is an equal condition. 

Before this PR:
```
mysql> desc select * from t where a >= 2 and a <= 2 order by b limit 2;
+----------------------------------+---------+-----------+--------------------------+---------------------------------------------+
| id                               | estRows | task      | access object            | operator info                               |
+----------------------------------+---------+-----------+--------------------------+---------------------------------------------+
| TopN_10                          | 0.01    | root      |                          | test.t.b, offset:0, count:2                 |
| └─IndexLookUp_27                 | 0.01    | root      |                          |                                             |
|   ├─IndexRangeScan_25(Build)     | 0.01    | cop[tikv] | table:t, index:idx(a, b) | range:[2,2], keep order:false, stats:pseudo |
|   └─TableRowIDScan_26(Probe)     | 0.01    | cop[tikv] | table:t                  | keep order:false, stats:pseudo              |
+----------------------------------+---------+-----------+--------------------------+---------------------------------------------+
4 rows in set (0.01 sec)
```

After this PR:
```
mysql> desc select * from t where a >= 2 and a <= 2 order by b limit 2;
+----------------------------------+---------+-----------+--------------------------+--------------------------------------------+
| id                               | estRows | task      | access object            | operator info                              |
+----------------------------------+---------+-----------+--------------------------+--------------------------------------------+
| Projection_32                    | 0.01    | root      |                          | test.t.a, test.t.b, test.t.c               |
| └─IndexLookUp_31                 | 0.01    | root      |                          | limit embedded(offset:0, count:2)          |
|   ├─Limit_30(Build)              | 0.01    | cop[tikv] |                          | offset:0, count:2                          |
|   │ └─IndexRangeScan_28          | 0.01    | cop[tikv] | table:t, index:idx(a, b) | range:[2,2], keep order:true, stats:pseudo |
|   └─TableRowIDScan_29(Probe)     | 0.01    | cop[tikv] | table:t                  | keep order:false, stats:pseudo             |
+----------------------------------+---------+-----------+--------------------------+--------------------------------------------+
5 rows in set (0.01 sec)

```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- choose a better plan when the range condition can convert to point

### CODE REVIEW

**Comment by Reminiscent** [comment]
@winoros @eurekaka PTAL. Thanks!

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -500,11 +500,18 @@ func (ds *DataSource) getIndexCandidate(path *util.AccessPath, prop *property.Ph
 	// When the prop is empty or `all` is false, `isMatchProp` is better to be `false` because
 	// it needs not to keep order for index scan.
 	if !prop.IsEmpty() && all {
+		eqCondCount := 0
+		for _, ran := range path.Ranges {
+			// Some expressions such as `a >= 2 && a <= 2` can also convert to a equal condition like `a = 2`.
+			if ran.IsPoint(ds.ctx.GetSessionVars().StmtCtx) {
+				eqCondCount++
```
Comment (eurekaka) [inline planner/core/find_best_task.go:507]: Would it be better to add this check / modification at `detachCNFCondAndBuildRangeForIndex`?

**Pull request closed without merging**